lib/linkerd2-opencensus/src/lib.rs

```diff
@@ -0,0 +1,186 @@
+    }
+
+    /// Attempt to read spans from the spans stream and write it to
+    /// the export streaming request.  Returns NotReady there isn't room in
+    /// the streaming request or if there isn't a span available in the spans
+    /// stream.
+    ///
+    /// Returns Ready if the receiver is fully consumed and may no longer be
+    /// used.
+    ///
+    /// An error is returned if the sender fails and must be discarded.
+    ///
+    /// Otherwise NotReady is returned.
+    fn poll_send_spans(
+        receiver: &mut S,
+        sender: &mut mpsc::Sender<ExportTraceServiceRequest>,
+        node: &mut Option<Node>,
+        max_batch_size: usize,
+    ) -> Poll<(), StreamError<S::Error>> {
+        try_ready!(sender.poll_ready().map_err(|_| StreamError::SenderLost));
+
+        let mut spans = Vec::new();
+        loop {
+            match receiver.poll() {
+                Ok(Async::NotReady) => {
+                    // If any spans have been collected send them, potentially consuming `node.
+                    Self::do_send(spans, sender, node)?;
+                    return Ok(Async::NotReady);
+                }
+                Ok(Async::Ready(Some(span))) => {
+                    spans.push(span);
+                    if spans.len() == max_batch_size {
+                        Self::do_send(spans, sender, node)?;
+                        // Because we've voluntarily stopped work due to a batch
+                        // size limitation, notify the task to be polled again
+                        // immediately.
+                        task::current().notify();
+                        return Ok(Async::NotReady);
+                    }
+                }
+                Ok(Async::Ready(None)) => {
+                    let _ = Self::do_send(spans, sender, node);
+                    // The span receiver stream completed, so signal completion.
+                    return Ok(Async::Ready(()));
+                }
+                Err(e) => {
+                    let _ = Self::do_send(spans, sender, node);
+                    return Err(StreamError::Receiver(e));
+                }
+            }
+        }
+    }
+}
+
+impl<T, S> Future for SpanExporter<T, S>
+where
+    T: GrpcService<BoxBody>,
+    S: Stream<Item = Span>,
+    S::Error: Into<Error>,
+{
+    type Item = ();
+    type Error = Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        loop {
+            self.state = match self.state {
+                State::Idle => {
+                    // If the request stream fails, all spans in this buffer
+                    // will be lost.  Therefore, we keep this buffer small to
+                    // minimize span loss.  Keeping this buffer small is okay
+                    // because we will exert backpressure and can buffer spans
+                    // elsewhere upstream.
+                    // TODO: Does this make sense?
+                    let (tx, rx) = mpsc::channel(1);
```

Comment:
As per the above comment, I don't think there's any reason to make this buffer larger.

Reply:
well, this is a good point... how many spans do we expect to be generated when transiting a single request through the proxy.. 2? it seems that it should be _at least_ as large as that, but using the self.max_batch_limit seems like a great idea to me!

Reply:
Consider that the `spans` Stream will be buffered upstream.  E.g.
```
        let (inbound_spans_tx, inbound_spans_rx) = mpsc::channel(100);
        let (outbound_spans_tx, outbound_spans_rx) = mpsc::channel(100);
        if let Some(trace_collector) = trace_collector_svc {
            let node = oc::Node {
                identifier: Some(oc::ProcessIdentifier {
                    host_name: hostname::get_hostname().unwrap_or_default(),
                    pid: 0,
                    start_timestamp: Some(start_time.into()),
                }),
                library_info: None,
                service_info: Some(oc::ServiceInfo {
                    name: "linkerd-proxy".to_string(),
                }),
                attributes: HashMap::new(),
            };
            let merged = inbound_spans_rx.select(outbound_spans_rx);
            let span_exporter = SpanExporter::new(trace_collector, node, merged);
            task::spawn(span_exporter);
        }
```

I'm having a hard time wrapping my head around how these two buffers interact with the batching of spans.  My initial thinking was to make only one buffer substantial and make the other one small and to use backpressure to push buffering into just one place, rather than needing to tune two separate buffer sizes.  But I think this interferes with our new buffering strategy.  Do both buffers need to be as large as the batch size in order for batching to work?  Since we need be able to move that many elements from one buffer to the next all in one task activation?

Reply:
Ah hah! of course. What do you think about renaming rx/tx to something like request_rx/tx so it's harder to get disoriented about this. This is, of course, the channel for sending messages on the client, and such there's basically no reason for this to be related to the batch size at all. It is more important that the _span_ channel is some factor of the max batch size to prevent spans from being needlessly dropped. We can consider this resolved from my POV.